# 1 Introducing Gephi

When you explore an unfamiliar complex network for the first time, it often
helps to perform a quick visual check of its structure before engaging in
expensive code writing. Sometimes, you can semi-automate even the network
construction itself (we are talking about small networks, with fewer than a
couple dozen nodes). You can perform many one-time construction, analysis, and conversion tasks with Gephi. Gephi is a free, Java-based, interactive CNA environment that runs
on all mainstream operating systems. 

In this lesson, you will learn:
- how to use Gephi and the data from network of nutrients and food types
- add attributes to its nodes and edges
- save the network as a presentation quality image
- You will also learn how to interchange network files between Gephi and NetworkX.

## 1.1 Worth 1,000 Words

Gephi is not a part of NetworkX, and you must install it separately. Luckily, the
installation process is [straightforward](https://gephi.org/users/install/) Here’s a summary of Gephi’s capabilities:
- Import existing networks in a variety of formats or create a new network.
- Edit a new or existing network by adding or removing nodes or edges.
- Change the size and color of node icons, the size and color of label font,
and the color and thickness of edges based on node and edge attributes.
- Calculate basic network measures: centralities, clustering coefficients,
path length distributions, connectedness, and modularity.
- Apply various layout engines to the network graph.
- Execute additional plug-ins.
- Export modified networks in a variety of formats.
- Save network visualization as a PNG, PDF, or SVG file.

Gephi is an excellent network construction and analysis tool, but it is interactive. The human user is its slowest component. Gephi cannot be programmed to
execute batches of tedious analysis tasks, vary parameters automatically, or
integrate with machine learning or predictive analytics software.



## 1.2 Import and Modify a Simple Network with Gephi

The following figure shows the standard main window of Gephi without any
loaded graphs.

<img width="600" src="https://drive.google.com/uc?export=view&id=12b2nJ1wzjURuhQqkGzQ0K6v5y4WI9yhy">

The main window contains three tabs: 
- **Overview** (for interactive network creation
and exploration)
- **Data Laboratory** (for text and numeric editing and research),
- **Preview** (for presentation-quality visualization).


We will start in the **Overview tab**, then briefly stop at the **Data Laboratory**, and finish in the **Preview**.


The default **Overview tab** has four empty windows. 

- The upper-left window, **Appearance**, is in charge of rendering. You will use it to control node and edge presentation properties. 
- The lower-left window, **Layout**, is behind the graph layout. 
- The central window, **Graph**, shows the sketch of the network. 
- The window on the right, with the tabs **Statistics and Filters**, is for network analysis
and filtering.

**To import an existing network file** prepared by NetworkX or any other CNA
software—in our case, **nutrients.csv** — open it with *File > Open (choose the Undirected graph type)*. 

Check if the number of detected edges and nodes makes sense. If the
import is successful, you will see an ugly black-and-white sketch of your
network in the Graph window. Don’t worry: we will make it look awesome by
the end of this section.



<img width="600" src="https://drive.google.com/uc?export=view&id=1o4NoBcI8LzjBpsrqC76efy5ftja5v_m9">


By default, Gephi treats networks as mathematical graphs and does not display
node labels. **Click the fat T button at the bottom of the Graph window**, and
the labels will show up, making the ugly sketch even uglier.

<img width="400" src="https://drive.google.com/uc?export=view&id=1D4RK-K80PJ7beLdpDyvp-RkSnjjB0i7C">


The original black color of the nodes is depressing. Before we go any further,
click the artist’s palette icon in the **Appearance window**, select the **Unique
tab**, and choose a fun color—say, light blue—for all the nodes. (Don’t forget
to click Apply!)

<img width="600" src="https://drive.google.com/uc?export=view&id=12wp9rN7xK0criSwVZRoIH7KoIlZl3uAP">


Suppose that at this point you realize that eggs are a valuable source of
selenium (they are) and you would like to add a connection between the
respective nodes. You remember that eggs are already on the network, but
you’re not sure about selenium, so you add it to the network, just in case. (It
is a mistake, but you will be able to correct it later.) Choose the **Node Pencil**
tool in the **Graph window**, adjust the color and size of the new node, if desired,
and click anywhere in the window. The position of the new node doesn’t
matter; the node will be moved around by the layout procedure.

<img width="600" src="https://drive.google.com/uc?export=view&id=1hnekUYK0Eorv7_F-6ju7Qcr5IghPtQnV">


Now, choose the **Edge Pencil** tool in the same window. Click the “eggs” node
and then the new nameless one. Congratulations, you added a new edge!


<img width="600" src="https://drive.google.com/uc?export=view&id=1WqIVCyjsk4AYAyr8sVk32WLJclvUfkBB">

Finally, take care of the anonymous node. It deserves a name, so you’re going
to name it “Se” for “selenium.” Click the Edit tab of the Appearance window,
then the node of interest. Edit the label name (the default value is “null
value”)—and realize that now you have two selenium nodes!


**Data Laboratory** (in addition to the **Edit tab**) is another place to look at the edges
and nodes under a microscope and inspect and modify their properties. The operation that you’re looking for—detection of duplicate nodes—is hidden in
the pull-down menu “More actions” at the top of the Data Table window. It is
called “Detect and merge node duplicates.” When you invoke the operation, Gephi
reports that it found one pair of duplicates, and offers to merge them into one
node. This procedure preserves all edges incident to the nodes. In particular, if
there is an edge connecting the duplicates, it becomes a self-loop edge.

Unfortunately, Gephi does not tell you which nodes are duplicates. You can
either trust the tool or manually find the duplicates in the table, select both
nodes, open the pop-up menu, and select “Merge nodes…” in it.

<img width="600" src="https://drive.google.com/uc?export=view&id=1jIm9ZhJ8KuKlFDth4ZfHd3ING96TgEbh">

Either way, you can go back to the Overview window, and you’ll see that there is now only
one Se node, and it is properly connected to the eggs. The network now looks
like the following figure. Node positions and fonts may differ, but surely the
new image is a vast improvement over the black-and-white sketch.

<img width="600" src="https://drive.google.com/uc?export=view&id=11QAUFxYPefL9Na313pxGEwJtm8dEhbil">


## 1.3 Explore the Network

Network exploration in Gephi goes hand in hand with selecting visual properties.
Let’s paint and resize the graph nodes based on some of their measures.

| Measure             | Meaning                                                                                                                                                                                                                                         |
|---------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Degree              | The number of immediate neighbors—adjacent nodes. <br>The degree is a non-negative integer number.<br> The larger the degree of a food item is, the more nutrients it provides. <br>The larger the degree of a nutrient is, the more food items provide it. |
| Community structure | Nodes form tightly knit groups called communities. <br>All foods and nutrients within a community serve some common purpose. <br>Each community has a unique integer identifier called modularity class.                                                |


Node degree is the simplest possible node measure. There is no need to calculate
it explicitly. To make node size proportional to the degree, click the
icon with concentric circles in the Appearance window, then on the Ranking
button. Select Degree from the “—Choose an attribute” pull-down menu.
Select node sizes that correspond to the smallest and largest degrees (2 and
60 are good choices). And don’t forget to click Apply. Can you see which nodes
have the highest degrees?

<img width="600" src="https://drive.google.com/uc?export=view&id=1VlTBoeTOASt-5XFKFtbHDk3NQio5G0cD">

Playing with size is fun; playing with color is more fun. Let’s paint the nodes
according to their modularity classes. To partition a network into communities, click the Run button next to the Modularity command in the Network Overview section of the Statistics window. Proceed by clicking OK and Close in the next two dialogs. In the end, you will
see a floating-point number next to Modularity. The number is a measure of
the quality of the decomposition from Outline Modularity-Based Communities. Return to the artist’s palette icon, then click the Partition button. Select Modularity Class from the “—Choose an attribute” pull-down menu. Don’t forget to click Apply.

<img width="600" src="https://drive.google.com/uc?export=view&id=1CgZ3YeAt-dJDB5lc_c4VVRwWvgnYM-OS">


## 1.4 Sketch the Network

You’re done with rough rendering, but the layout is still awful. Let’s turn
our attention to the lower-left corner of Gephi. Select your favorite layout from
the “Choose a layout” pull-down menu. When a network is large (500 or more
nodes), the Fruchterman-Reingold layout is usually the most efficient. For
smaller networks, the ForceAtlas 2 layout, with some tweaking, works marvels.
To make things easier, here’s a tip: set the scaling to 100.0 (to place
nodes reasonably far apart) and check the Prevent Overlap box. Run the tool
for a while. You’ll notice that after a couple of seconds, the nodes settle at
their new positions, but the graph as a whole may continue drifting, rotating,
or both.

<img width="600" src="https://drive.google.com/uc?export=view&id=1BazbvKbE-ygbtPeOz98SeerCIYoKxa29">

The last step is to adjust the labels, because surely some of them don’t mind
their manners and sit on top of each other. Select Label Adjust from the
“Choose a layout” pull-down menu and run the tool for a couple of seconds.
This layout engine distorts the original Fruchterman-Reingold but makes
sure that neither nodes nor labels overlap. Hopefully, your network layout
resembles the following figure.

<img width="600" src="https://drive.google.com/uc?export=view&id=1cbA550qLPoQYyaRpAobpSN4MjVEyuTiQ">


The Graph window still shows a sketch, but this is a high-quality sketch
that nicely displays the structure of the network of foods and nutrients. But it would look much better when rendered at high resolution and converted to a presentationquality
image. (Save the project via File > Save As… into a .gephi file to avoid
data loss if Gephi crashes!)

## 1.5 Prepare a Presentation-Quality Image

The final painting of the network takes place in the Preview tab. The tab has
two windows: Preview Settings (to control the fine-level rendering engine)
and Preview (to see the rendering results). When you switch to this tab, the
Preview window will be empty. Click the Refresh button. Your network will
be drawn using the default settings, with curved edges and no node labels
(see the following figure). Don’t worry—the labels have not been lost; they
have been turned off.

<img width="600" src="https://drive.google.com/uc?export=view&id=1pTDvb2XDu_AkhYR7W9Dmq15ci62iF9Ee">

Gephi supports a variety of preset renderers. The default one is usually not
the best one. I recommend using the **Text outline** preset configuration with
some extra tweaking:
- Unclick “Proportional size.”
- Increase the Font size 
- Increase edge Thickness 
- Set edge Opacity at 75.0.
- Optionally, unclick the Curved box.

Click the Refresh button again.

Once you like the rendering results, you can export the graph into a graphics
file. Gephi provides exporters to SVG (editable vector format), PDF (editable
vector format), and PNG (non-editable raster format).

<img width="600" src="https://drive.google.com/uc?export=view&id=1d5DutZD5ozrThHqoyDXoLa3XO3CdKFCJ">

# 2 Combine Gephi and NetworkX

There is no implicit integration between Gephi and NetworkX. However, you can use
Gephi graph file exporter (File > Export > Graph file…) to save your network into
a file that could be imported by NetworkX. **GraphML** is the preferred interchange
format because it preserves all calculated measures (such as centralities and
modularity classes) as node attributes. This way you could use Gephi to perform
a quick-and-dirty interactive analysis of a network and save it into a **.graphml** file
for further processing. 

Export the previous network to the file **nutrients.graphml**.

<img width="600" src="https://drive.google.com/uc?export=view&id=1x_U2Kn1L0xE2MQWKTFIqsE4HCPhtzX5W">


You just learned one more way to build and analyze networks by hand, and you
now understand that analyzing one not-so-large, pre-packaged complex network
with Gephi is a pleasure. You also understand that constructing and analyzing
many large networks by hand is hard, error-prone, and most of the time infeasible.
You are now ready for the first full-Python case study.

**Exercise (reading)**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">

1. Use the function **nx.read_graphml()** to convert nutrient.graphml to a networkx object named **g**.
2. Find the number of nodes and egdes in the graph and compare with the values found in Gephi.
3. Verify the  attributes of nodes and edges.

In [0]:
# put your code here

**Exercise (visualizing)**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">

1. Print the node A and see the template of output.
```python
g.nodes["A"]
```
2. Use a dictionary comprehension to iterate over g and create a dictionary like that:
```python
{node: [g.nodes[node]['x'], g.nodes[node]['y']]}
```
  - Store the dictionary to the variable **pos**.
3. Use a dictionary comprehension to iterate over g and create a dictionary like that:
```python
{node: g.nodes[node]['label']}
```
  - Store the dictionary to the variable **labels**.
4. Use a list comprehension to iterate over g and create a list like that:
```python
[g.nodes[node]["size"]*10]
```
  - Store the list to the variable **node_size**
5. Use a list comprehension to iterate over g and create a list like that:
```python
from matplotlib.colors import rgb2hex
[rgb2hex((g.nodes[node]['r']/255,
                  g.nodes[node]['g']/255,
                  g.nodes[node]['b']/255)
                )]
 ```
  - Store the list to the variable **node_color**
6. Plot the graph g using the function **nx.draw_networkx()** with the parameters: g, pos, labels, node_size, node_color and ax.
```python
fig, ax = plt.subplots(figsize=(10,10))
```


In [0]:
# put your code here

# 3 Case Study: constructing a network of wikipedia pages

## 3.1 Introduction

So far we have learned two ways of constructing a complex network:
- from a CSV file 
- and further in Gephi

What is hard for small networks may be impossible for medium-to-large scale networks; it may be impossible even for small networks if you must repeat the analysis many
times. The case study in this lesson shows you **how to construct a large
network in an easy way**:  
- automatically collecting node and edge data from the Internet.

The other goal of this study (aside from mastering new network construction
techniques) is quite pragmatic. 
- Wouldn’t you want to know where the **complex
network analysis** fits in the context of other subjects and disciplines? 

An answer to this question is near at hand: [on Wikipedia](https://en.wikipedia.org/wiki/Complex_network).


Let’s start with the Wikipedia page about complex networks—the seed page.
(Unfortunately, there is no page on complex network analysis itself.) The page
body has external links and links to other Wikipedia pages. Those other pages
presumably are somewhat related to complex networks, or else why would
the Wikipedia editors provide them?
To build a network out of the seed page and other relevant pages, **let’s treat
the pages (and the respective Wikipedia subjects) as the network nodes and
the links between the pages as the network edges**. You will use snowball
sampling (a breadth-first search or [BFS algorithm](https://en.wikipedia.org/wiki/Breadth-first_search)) to discover all the nodes and edges of interest.

<img width="200" src="https://drive.google.com/uc?export=view&id=1t9ZXB5Q-wRt_qWbf8eU_1RiTwVGaStHU">

As a result, you will have a network of all pages related to complex networks
and hopefully, you will make some conclusions about it.

## 3.2 Get the Data, Build the Network

The first half of the project script consists of:

1. the initialization prologue
2. a heavy-duty loop that retrieves the Wikipedia pages and simultaneously builds the network of nodes and edges.

Let’s first import all necessary modules. We will need:

- the module **wikipedia** for fetching and exploring Wikipedia pages
- the operator **itemgetter** for sorting a list of tuples
- and, naturally, **networkx** itself.

To target the **snowballing process**, define the constant **SEED, the name of the
starting page**. 

> As a side note, by changing the name of the seed page, you can
apply this analysis to any other subject on Wikipedia.


Last but not least, when you start the snowballing, you will eventually (and
quite soon) bump into the pages describing ISBN and ISSN numbers, the
arXiv, PubMed, and the like. Almost all other Wikipedia pages refer to one or
more of those pages. This hyper-connectedness transforms any network into
a collection of almost perfect gigantic stars, making all Wikipedia-based networks
look similar. **To avoid the stardom syndrome**, treat the known “star”
pages as stop words in information retrieval—in other words, ignore any links
to them. 

Constructing the **black list of stop words**, STOPS, is a matter of trial
and error. We put thirteen subjects on it; you may want to add more when you
come across other “stars.” We also excluded pages whose names begin with
**"List of"**, because they are simply lists of other subjects.

In [0]:
!pip install wikipedia

In [0]:
from operator import itemgetter
import networkx as nx
import wikipedia

In [0]:
SEED = "Complex network".title()
STOPS = ("International Standard Serial Number",
         "International Standard Book Number",
         "National Diet Library",
         "International Standard Name Identifier",
         "International Standard Book Number (Identifier)",
         "Pubmed Identifier", 
         "Pubmed Central",
         "Digital Object Identifier", 
         "Arxiv",
         "Proc Natl Acad Sci Usa", 
         "Bibcode",
         "Library Of Congress Control Number", 
         "Jstor")

The next code fragment deals with setting up the **snowballing process**. A
breadth-first search, or BFS (sometimes known to computer programmers
as a snowballing algorithm), must remember which **pages have been already
processed** and which have been **discovered but not yet processed**. 
- the former are stored in the set **done_set**; 
- the latter, in the list **todo_lst** and **todo_set**. 

You need two data structures for the unprocessed pages because you want to
know whether a page has been already recorded (an unordered lookup, *todo_set*) and
which page is the next to be processed (an ordered lookup, *todo_lst*). 


Snowballing an extensive network—and Wikipedia with 5,845,876 articles in
the English segment alone can produce a huge network!—takes considerable
time.

<img width="300" src="https://drive.google.com/uc?export=view&id=1tX6HB0R0Pieal5AQROZ7dyaZzzN7wgwZ">

Suppose you start with one seed node, and let’s say it has $N\approx 100$ neighbors.
Each of them has $N$ neighbors, too, to the total of $\approx N+N×N$ nodes. The
third round of discovery adds $\approx N×N×N$ more nodes. The time to shave each
next layer of nodes grows exponentially. For this exercise, let’s process only
the seed node itself and its immediate neighbors (layers 0 and 1). Processing
layer 2 is still feasible, but layer 3 requires $N×N×N×N\approx 10^8$ page downloads. To keep track of the distance from the currently processed node to the seed, store both the layer to which a node belongs and the node name together as a tuple on the **todo_lst** list.

In [0]:
todo_lst = [(0, SEED)] # The SEED is in the layer 0
todo_set = set(SEED) # The SEED itself
done_set = set() # Nothing is done yet

**The output of the exercise is a NetworkX graph**. The next fragment will create
an empty directed graph that will later absorb discovered nodes and edges.

> We choose a **directed graph** because the edges that represent HTML links
are naturally directed: a link from page A to page B does not imply a reciprocal
link. 

The same fragment primes the algorithm by extracting the first “to-do” item
(both its layer and page name) from the namesake list.



In [0]:
g = nx.DiGraph()
layer, page = todo_lst[0]

It may take a fraction of a second to execute the first five lines of the script.
It may take the whole next year or longer to finish the next twenty lines
because they contain the main collection/construction loop of the project.

In [0]:
%%time
while layer < 2:
  # Remove the name page of the current page from the todo_lst, 
  # and add it to the set of processed pages. 
  # If the script encounters this page again, it will skip over it.
  del todo_lst[0]
  done_set.add(page)
  
  # Show progress
  print(layer, page) 
  
  # Attempt to download the selected page.
  try:
    wiki = wikipedia.page(page)
  except:
    layer, page = todo_lst[0]
    print("Could not load", page)
    continue
  
  for link in wiki.links:
    link = link.title()
    if link not in STOPS and not link.startswith("List Of"):
      if link not in todo_set and link not in done_set:
        todo_lst.append((layer + 1, link))
        todo_set.add(link)
      g.add_edge(page, link)
  layer, page = todo_lst[0]

In [0]:
print("{} nodes, {} edges".format(len(g), nx.number_of_edges(g)))

11738 nodes, 22716 edges


The network of interest is now in the variable g. But it is “dirty”: inaccurate, incomplete, and erroneous.

## 3.3 Eliminate Duplicates

Many Wikipedia pages exist under two or more names. For example, there
are pages about **Complex Network** and **Complex Networks**. The latter redirects
to the former, but NetworkX does not know about the redirection.

Accurately merging all duplicate nodes involves **natural language processing
(NLP) tools that are outside of the scope of this course**. It may suffice to join
only those nodes that differ by the presence/absence of the letter s at the end
or a hyphen in the middle.

Start removing self-loops (pages referring to themselves). The loops don’t change
the network properties but affect the correctness of duplicate node elimination.
Now, you need a list of at least some duplicate nodes. You can build it by
looking at each node in g and checking if a node with the same name, but
with an s at the end, is also in g. 

Pass each pair of duplicated node names to the function **nx.contracted_nodes(g,u,v)** that merges node v into node u in the graph g. The function reassigns all edges previously incident to v, to u. If you don’t pass the option **self_loops=False**, the function converts an edge from v to u (if any) to a self-loop.

In [0]:
# remove self loops
g.remove_edges_from(g.selfloop_edges())

# identify duplicates like that: 'network' and 'networks'
duplicates = [(node, node + "s") 
              for node in g if node + "s" in g
             ]

for dup in duplicates:
  # *dup is a technique named 'unpacking'
  g = nx.contracted_nodes(g, *dup, self_loops=False)

duplicates = [(x, y) for x, y in 
              [(node, node.replace("-", " ")) for node in g]
                if x != y and y in g]

for dup in duplicates:
  g = nx.contracted_nodes(g, *dup, self_loops=False)

# nx.contracted creates a new node attribute called contraction
# the value of the attribute is a dictionary, but GraphML
# does not support dictionary attributes
nx.set_node_attributes(g, 0,"contraction")

In [0]:
print("{} nodes, {} edges".format(len(g), nx.number_of_edges(g)))

11613 nodes, 21574 edges


## 3.4 Truncate the network

Why did you go through all those Wikipedia troubles? First, to construct a
network of subjects related to complex networks—and here it is. Second, to
find other significant topics related to complex networks. **But what is the
measure of significance?**


You will discover a variety of network measures latter in this course. For now, let’s concentrate on a **node indegree**—the number of edges directed into the node. (In the same spirit, the number of edges directed out of the node is called **outdegree**.) The indegree of a node equals the number of HTML links pointing to the respective page. **If a page has a lot of links to it, the topic of the page must be significant.**


The choice of indegree as a yardstick of significance incidentally makes it
possible to shrink the graph size by almost 75 percent. The extracted graph
has 11,613 nodes and 21,574 edges—an average of 1.8 edges per node. **Most
of the nodes have only one connection**. (Interestingly, there are no isolated
nodes with no connection in the graph. Even if they exist, you will not find
them because of the way snowballing works.) **You can remove all nodes with
only one incident edge to make the network more compact and less hairy
without hurting the final results**.

In [0]:
# filter nodes with degree greater than or equal to 2
core = [node for node, deg in dict(g.degree()).items() if deg >= 2]

# select a subgraph with 'core' nodes
gsub = nx.subgraph(g, core)

print("{} nodes, {} edges".format(len(gsub), nx.number_of_edges(gsub)))

nx.write_graphml(gsub, "cna.graphml")

3199 nodes, 13160 edges


As you can see, the following code fragment safely removes 72 percent of the
nodes and 39 percent of the edges, raising the average number of edges per
node to 4.11.

## 3.5 Explore the network

The following figure is a Gephi rendering of **gsub**. Node
and label font sizes represent the indegrees (a filter was added in gephi just to show only node with indegree greater than 50). The most in-connected, most significant
nodes are in the upper-left corner of the network. What are they?

<img width="600" src="https://drive.google.com/uc?export=view&id=1Cemnpe6yBttyZ8Vw2i9ifMiWx9psIV7C">


The last code fragment of the exercise efficiently calculates the answer by
calling the method **gsub.in_degree()**. The method (and its sister method **gsub.out_degree()**)
are very similar to **gsub.degree()** except that they report different edge counts in the form of objects InDegreeView and OutDegreeView, respectively.

```python
top_indegree = sorted(dict(gsub.in_degree()).items(),
                      reverse=True, key=itemgetter(1))[:100]
print("\n".join(map(lambda t: "{} {}".format(*reversed(t)), top_indegree)))
```

In [0]:
top_indegree = sorted(dict(gsub.in_degree()).items(),
                      reverse=True, key=itemgetter(1))[:100]
print("\n".join(map(lambda t: "{} {}".format(*reversed(t)), top_indegree)))

65 Graph (Discrete Mathematics)
63 Vertex (Graph Theory)
56 Directed Graph
54 Social Network
50 Network Theory
50 Degree (Graph Theory)
47 Edge (Graph Theory)
47 Graph Drawing
46 Adjacency Matrix
46 Bipartite Graph
45 Graph (Abstract Data Type)
45 Graph Theory
45 Complete Graph
44 Cycle (Graph Theory)
44 Network Science
44 Incidence Matrix
43 Loop (Graph Theory)
43 Complex Network
43 Social Network Analysis Software
43 Social Capital
42 Scale-Free Network
42 Small-World Network
42 Centrality
42 Preferential Attachment
42 Complex Contagion
42 Path (Graph Theory)
41 Hypergraph
41 Adjacency List
40 Distance (Graph Theory)
40 Random Graph
40 Clustering Coefficient
39 Flow Network
39 Betweenness Centrality
39 Clique (Graph Theory)
39 Multigraph
39 Connected Component (Graph Theory)
38 Percolation Theory
38 Semantic Network
38 Watts And Strogatz Model
38 Citeseerx
38 Computer Network
37 Combinatorial Optimization
37 Transitive Relation
37 Degree Distribution
37 Barabási–Albert Model
37 Cut (

**Exercise**

<img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ">


- This section presented a complete complex network construction case study,
starting from the raw data in the form of HTML pages, all the way to an analyzable
annotated network graph and a simple exploratory exercise. This is
a good foundation for more systematic complex network studies. Now is your turn, choice a topic of your interest and reproduce yourself a complet complex network case study. 

In [0]:
# put your results here